### PR TITLE
#14138 - Improve Logging and Tracing

### PR DESCRIPTION
**Pull request by billy-editiano**
## Changes
- LoggerService structure 
  - add LoggerService and its creator
  - add handleRunConsumer (a dummy command action) to demonstrate the usage of the logger
- New Dependencies
  - [`cls-hooked`](https://github.com/jeff-lewis/cls-hooked) to provide a context and jobId
  - [`winston-daily-rotate-file`](https://github.com/winstonjs/winston-daily-rotate-file) winston transport for daily rotate file

### CODE REVIEW

**Comment by tonoman3g** [comment]
@billy-editiano can you attach screenshot or log file example?

File: src/index.ts [hunk 1]
```diff
@@ -51,6 +55,8 @@ program
   .description('Applies tags to projects based on the given config file')
   .action(handleApplyTags);
 
+program.command(`run-consumer <configFile>`).description('Run Consumer').action(handleRunConsumer);
```
Comment (tonoman3g) [inline src/index.ts:58]: can we separate this branch into two PRs

1. PR to master only contains the logger improvement
2. PR to rex contains example of run-consumer handler

this way, the master branch will have benefit of the logging improvement

File: src/handlers/create-project.handler.ts [hunk 2]
```diff
@@ -6,13 +6,20 @@ import { getProjectCreationValidators } from '../config/schema/validator';
...
 import { getLabelSetsFromDirectory } from '../utils/labelset';
 import { sleep } from '../utils/sleep';
 import { ScriptAction } from './constants';
 
 export async function handleCreateProject(projectName: string, configFile: string) {
   const cwd = process.cwd();
+
+  getLoggerService().registerResolver(() => {
+    return {
+      command: 'create-project',
```
Comment (tonoman3g) [inline src/handlers/create-project.handler.ts:19]: where will this resolver.command name be used for?
will it be printed in the log?

File: src/handlers/run-consumer.ts [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+const random = () => {
+  return Math.floor(Math.random() * 5) + 1;
+};
+
+const contextKey = 'jobId';
+
+// simulate consumer
+function consumeDataFromDatabase(onJob: (jobId: number) => void | Promise<void>) {
+  setInterval(() => {
+    const randomJobId = random();
```
Comment (tonoman3g) [inline src/handlers/run-consumer.ts:13]: oh, the request id will only works for run-consumer handler ya?

can we do this for all handler?

so basically every api call from robosaur will have the request id

File: src/handlers/create-project.handler.ts (hunk 2 above)
Comment (billy-editiano) [reply src/handlers/create-project.handler.ts:19]: yes, i think this is good for tracing so i add this "command" field

File: src/index.ts (hunk 1 above)
Comment (billy-editiano) [reply src/index.ts:58]: ahh sure, good idea.. okay i'll start working on it

**Comment by billy-editiano** [comment]
Hi ka @tonoman3g this is the [PR](https://github.com/datasaur-ai/robosaur/pull/59) to master (i created another branch to avoid conflicts)

File: src/handlers/run-consumer.ts [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+  setTimeout(() => {
+    getLogger().info('job completed');
+  }, 120);
+}
+
+// simulate get job from database
+export async function _handleRunConsumer(process: ProcessJob<unknown[]>, configFile: string) {
+  getLogger().info('Begin running consumer', { configFile });
+  setInterval(() => {
+    const job: Job = { jobId: `${random()}`, payload: 'some-payload-data' };
```
Comment (tonoman3g) [inline src/handlers/run-consumer.ts:27]: change jobId to `trace-id`

**Pull request merged**